### ISSUE TITLE
Support typing.NewType

### ISSUE THREAD

**Issue by mcaramma** [jcrist/msgspec#250]
Hi!
Is it possible to add support for `NewType` ?

Example:
```python
Cuid = Annotated[NewType("Cuid", str), Meta(min_length=25, max_length=25, pattern=r"^c[0-9a-z]+$")]
```
It can be argued that this code achieve the same
```python
Cuid = Annotated[str, Meta(min_length=25, max_length=25, pattern=r"^c[0-9a-z]+$")]
```
But semantically are very different, both static (`mypy`) and dynamic (`beartype`) type checking will not allow
```python
@beartype
def foo(cuid: Cuid):
    pass

foo("cjld2cjxh0000qzrmn831i7rn")  # should raise error
```
if `Cuid` is defined with `NewType` ... unfortunately `msgspec` doesn't support it.

A naive approximation might be something like
```python
def NewTyped(name: str, type: Type, *args):
    return msgspec.defstruct(
        name,
        [("v", Annotated[tuple([type, *args])] if args else type)],
    )

Cuid = NewTyped("Cuid", str, Meta(min_length=25, max_length=25, pattern=r"^c[0-9a-z]+$"))
```
But this is sub-optimal, because a `Cuid` instance is not represented by a string (`{"v": "cjld2cjxh0000qzrmn831i7rn"}`).

**Comment by jcrist** [comment]
Can you say more about your use case? Treating a `NewType` as the same as its base type in msgspec would be very straightforward to do (rather than allowing specializing decoding on the `NewType`). But I'm not sure how common this use case is. What does using a `NewType` with msgspec without specialization allow you to do here?

**Comment by jcrist** [comment]
See #251 for a PR implementing what you're asking for. Still not 100% committed to it, but it's easy to support if this behavior makes sense.

**Comment by mcaramma** [comment]
Wow, that was really quick ... thanks!

To answer about a specific use case
```python

Filter = Annotated[str, Meta(...)]
SanitizedFilter = Annotated[NewType("Filter"), Meta(...)]

class Query(msgspec.Struct):
    filter: SanitizedFilter

def sanitize_filter(filter: Filter) -> SanitizedFilter:
    [...]

async def query(filter: Filter) -> Result:
    sanitized_filter = sanitize_filter(filter)
    q = Query(filter=sanitized_filter)
    return await remote_service(serialize(q))
```
By making SanitizedFilter a NewType, make sure that only properly sanitized `filter` are operated upon, eliminating the chance of unwanted behaviours.

Otherwise a sloppy use of `Query(filter=filter)` will pass unnoticed.

**Comment by jcrist** [comment]
Makes sense to me, thanks!

This has been resolved by #251. Until the next release you'll have to work from the `main` branch. If you have a c compiler installed, you can install `msgspec` via:

```
pip install git+https://github.com/jcrist/msgspec.git
```

**Comment by mcaramma** [comment]
@jcrist thanks a lot!

**Comment by jcrist** [comment]
No problem! This has now been released as part of 0.12.0. Please open a new issue if you run into any other problems!
